### ISSUE TITLE
[Llama3] Create a beautiful generation interface

### ISSUE THREAD

**Issue by cglagovichTT** [tenstorrent/tt-metal#15058]
The Llama3 shared codebase demo currently handles prefill input prep, looped prefill, decode input prep, decode trace capture, and decode trace execution.

The Llama3 demo should be refactored to use a generator interface which takes care of these details. This will 
- clean up the demo
- make it easier to reason about tracing
- make vLLM integration possible for the text and vision models

The `LlamaGenerator` class will support both text and vision inference. 

## Prototype
I would like for this `LlamaGenerator` class to eventually be general to any LLM that we bring up. For that reason, the model itself will be required to implement a few functions which `LlamaGenerator` uses. These are the functions that the model must implement:
### Model interface
```python

    def prepare_inputs_prefill(self, *inputs):
        """
        Inputs are torch tensors or python types. This function returns ttnn 
        tensors on device.
        """
       

    def prepare_inputs_decode(self, *inputs):
        """
        Inputs are torch tensors or python types. This function returns ttnn
        tensors on device.
        Its implementation can take advantage of a few other functions which the 
        model must implement.
        """
        host_inputs = self.prepare_decode_inputs_host(*inputs)
        device_inputs = copy_host_to_device(host_inputs) # Helper function
        transformed_device_inputs = self.transform_decode_inputs_device(*device_inputs)
        return transformed_device_inputs

    def prepare_decode_inputs_host(self, *inputs):
        """
        Inputs are torch tensors or python types. Outputs are ttnn tensors on host.
        """

    def transform_decode_inputs_device(self, *device_inputs):
        """
        Inputs are ttnn tensors on device. This function applies any on-device 
        transformations which should happen before forward decode.
        For example: tilize, reshape, shard.
        Return transformed device tensors
        """

    def process_output_prefill(self, tt_out):
        """
        Input is ttnn device tensor of logits. Output is torch logits tensor
        """

    def process_output_decode(self, tt_out):
        """
        Input is ttnn device tensor of logits. Output is torch logits tensor
        """

    def ttnn_prefill_forward(
        self,
        device_inputs,
        other_args
    ):
        """
        This method will take device tensors and any other args to run forward.
        It returns ttnn device tensors.
        """

    def ttnn_decode_forward(
        self,
        device_inputs,
        other_args
    ):
        """
        This method will take device tensors and any other args to run forward.
        It returns ttnn device tensors.
        """
```

### Generator interface
Given those methods that the model supports, the Generator will compose them in order to enable inferece.

```python

class LlamaGenerator:
    def __init__(self, model, model_args, mesh_device, vllm=False, tokenizer=None, formatter=None):
        self.model = model
        self.model_args = model_args
        self.mesh_device = mesh_device
        self.vllm = vllm
        self.tokenizer = tokenizer
        self.formatter = formatter

    def prefill_forward_single_user(self, tokens, user_id, *other_inputs):

        tt_prefill_inputs = self.model.prepare_inputs_prefill(tokens, *other_inputs)

        tt_logits = self.model.ttnn_prefill_forward(
            *tt_prefill_inputs, user_id=user_id
        )

        logits = self.model.process_output_prefill(tt_logits)

        return logits

    def prefill_forward(self, tokens, *other_inputs):
        """
        Iterates over batch and calls prefill_forward_single_user
        """

    def decode_forward(
        self,
        position_id,
        tokens,
        page_table,
    ):

        tt_decode_inputs = self.model.prepare_inputs_decode(
            tokens, position_id, page_table
        )

        tt_logits = self.model.ttnn_decode_forward(*tt_decode_inputs)

        logits = self.model.process_output_decode(tt_logits)
        return logits

    def capture_trace(
        self,
        position_id,
        tokens,
    ):
        # Compile run
        tt_decode_inputs = self.model.prepare_inputs_decode(
            tokens, position_id=position_id
        )

        tt_logits_rm = self.model.ttnn_decode_forward(*tt_decode_inputs)

        # Get inputs ready for trace run
        host_decode_inputs = self.model.prepare_decode_inputs_host(tokens, position_id)

        tt_decode_inputs = self.model.copy_host_to_device(host_decode_inputs)

        trace_id = ttnn.begin_trace_capture(self.mesh_device, cq_id=0)
        # Do on-device transformations of inputs before forward
        transformed_decode_inputs = self.model.transform_decode_inputs_device(tt_decode_inputs)

        tt_logits_rm = self.model.ttnn_decode_forward(*transformed_decode_inputs)
        ttnn.end_trace_capture(self.mesh_device, trace_id, cq_id=0)

        return trace_id, tt_logits_rm, *tt_decode_inputs

    def decode_forward_trace(
        self,
        position_id,
        tokens,
        trace_id,
        *tt_decode_inputs,
        tt_logits_rm,
    ):
        host_decode_inputs = self.model.prepare_decode_inputs_host(
            tokens, position_id=position_id
        )

        self.model.copy_host_to_device(
            host_tensors=host_decode_inputs,
            device_tensors=tt_decode_inputs,
        )

        ttnn.execute_trace(self.mesh_device, trace_id, cq_id=0, blocking=False)
        logits = self.model.process_output_decode(tt_logits_rm)
        return logits

    def easy_trace(
        self,
        position_id,
        tokens,
    ):
        """
        Tracing is easy! Just call this method and we'll handle tracing for you.
        """
        if not hasattr(self, "trace_id"):
            trace_id, tt_logits_rm, *tt_decode_inputs = self.capture_trace(
                position_id,
                tokens,
            )
            self.trace_id = trace_id
            self.trace_inputs = tt_decode_inputs
            self.trace_outputs = tt_logits_rm

        return self.decode_forward_trace(
            position_id,
            tokens,
            self.trace_id,
            self.trace_inputs,
            self.trace_outputs,
        )

```

Some other principles:
- `Generator` should be responsible for padding and unpadding. For example, prefill sequence lengths must be a multiple of 1024 or a power of 2. Generator should pad the input and unpad the output
- `rope` should be a part of the model, not the demo code
- demo code should exclusively deal with torch tensors or python lists, never ttnn tensors

**Comment by cglagovichTT** [comment]
@mtairum @yieldthought @sraizada-tt lmk if you have feedback on the proposal. We can debate naming conventions, the responsibilities of functions, etc.

**Comment by sraizada-tt** [comment]
looks good to me
